src/languages/json-source-code.js

```diff
@@ -98,6 +117,135 @@ export class JSONSourceCode extends TextSourceCodeBase {
+			return this.getText(comment).slice(2, -2); // strip leading `/*` and trailing `*/`
+		}
+
+		throw new Error(`Unexpected comment type '${comment.type}'`);
+	}
+
+	/**
+	 * Returns an array of all inline configuration nodes found in the
+	 * source code.
+	 * @returns {Array<JSONToken>} An array of all inline configuration nodes.
+	 */
+	getInlineConfigNodes() {
+		if (!this.#inlineConfigComments) {
+			this.#inlineConfigComments = this.comments.filter(comment =>
+				INLINE_CONFIG.test(this.#getCommentValue(comment)),
+			);
+		}
+
+		return this.#inlineConfigComments;
+	}
+
+	/**
+	 * Returns directives that enable or disable rules along with any problems
+	 * encountered while parsing the directives.
+	 * @returns {{problems:Array<FileProblem>,directives:Array<Directive>}} Information
+	 *      that ESLint needs to further process the directives.
+	 */
+	getDisableDirectives() {
+		const problems = [];
+		const directives = [];
+
+		this.getInlineConfigNodes().forEach(comment => {
+			const { label, value, justification } =
+				commentParser.parseDirective(this.#getCommentValue(comment));
+
+			// `eslint-disable-line` directives are not allowed to span multiple lines as it would be confusing to which lines they apply
+			if (
+				label === "eslint-disable-line" &&
+				comment.loc.start.line !== comment.loc.end.line
+			) {
+				const message = `${label} comment should not span multiple lines.`;
+
+				problems.push({
+					ruleId: null,
+					message,
+					loc: comment.loc,
+				});
+				return;
+			}
+
+			switch (label) {
+				case "eslint-disable":
+				case "eslint-enable":
+				case "eslint-disable-next-line":
+				case "eslint-disable-line": {
+					const directiveType = label.slice("eslint-".length);
+
+					directives.push(
+						new Directive({
+							type: /** @type {DirectiveType} */ (directiveType),
+							node: comment,
+							value,
+							justification,
+						}),
+					);
+				}
+
+				// no default
+			}
+		});
+
+		return { problems, directives };
+	}
+
+	/**
+	 * Returns inline rule configurations along with any problems
+	 * encountered while parsing the configurations.
+	 * @returns {{problems:Array<FileProblem>,configs:Array<{config:{rules:RulesConfig},loc:SourceLocation}>}} Information
+	 *      that ESLint needs to further process the rule configurations.
+	 */
+	applyInlineConfig() {
+		const problems = [];
+		const configs = [];
+
+		this.getInlineConfigNodes().forEach(comment => {
+			const { label, value } = commentParser.parseDirective(
+				this.#getCommentValue(comment),
+			);
+
+			if (label === "eslint") {
+				const parseResult = commentParser.parseJSONLikeConfig(value);
+
+				if (parseResult.ok) {
+					configs.push({
+						config: {
+							rules: parseResult.config,
+						},
+						loc: comment.loc,
+					});
+
+					// Explicit check needed for TypeScript
+				} else if (parseResult.ok === false) {
+					problems.push({
+						ruleId: null,
+						message: parseResult.error.message,
```

Comment:
I inserted `if (parseResult.ok === false)` here because I was getting this TypeScript error without it (with just `else`):

```
error TS2339: Property 'error' does not exist on type '{ ok: true; config: RulesConfig; } | { ok: false; error: { message: string; }; }'.
  Property 'error' does not exist on type '{ ok: true; config: RulesConfig; }'.

229       message: parseResult.error.message,
```

It seems that TypeScript doesn't infer that in `else` of `if (parseResult.ok)`, `parseResult` must be `{ ok: false; error: { message: string; }; }`. Please let me know if there's a better solution.

Reply:
I think this is happening because we don't use the [`strictNullChecks`](https://www.typescriptlang.org/tsconfig/#strictNullChecks) option in tsconfig.json. Without that option, TypeScript assumes that the `ok` field in `parseResult` could be possibly nullish in both alternatives of the union type, so it cannot infer that the `error` property exists from the fact that `ok` is falsy.

You could add a JSDoc annotation like `/** @type {*} */` above the line where `parseResult` is declared to disable type checking in this case, but maybe there's an even better solution.

Reply:
Another way to do this is with a type guard, something like:

```js
/**
 * @param {ParseResult} result
 * @returns {result is OkParseResult}
 */
function isOkParseResult(result) {
    return result.ok;
}
```

Then you can use `isOkParseResult()` and TypeScript can know that you're positively identifying the type.

Reply:
Okay, I fixed this with casting.